### PR TITLE
fix NCCL error with non-consecutive trainer gpus

### LINKED ISSUE #8139: DDP and gpus=[1,2,3] causes NCCL error
## 🐛 Bug

On a machine with 4 GPUs, using DDP and `gpus=[1,2,3]` results in an NCCL error:

```
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1614378062065/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled system error, NCCL version 2.7.8
ncclSystemError: System call (socket, malloc, munmap, etc) failed.
```

The same error occurs for `[1,2]` and `[1,3]`.

For `[0,2]`, `[0,3]`, `[0,1,3]`, and `[0,2,3]`, the execution hangs. This may be related to #5264, although I'm using pytorch 1.8.

For `[0,1,2,3]`, `[0,1,2]`, `[0,1]`, `[2,3]`, or any single GPU, everything runs fine.

(I.e. continuous indices run fine. Failure occurs when there is a gap in indices: hanging if GPU 0 is included, otherwise giving the NCCL error.)

### To Reproduce

On a machine with 4 GPUs:

```
conda create -n bug_env
conda activate bug_env
conda install pytorch==1.8.0 torchvision==0.9.0 torchaudio==0.8.0 cudatoolkit=10.2 -c pytorch
pip install pytorch-lightning@git+https://github.com/PytorchLightning/pytorch-lightning.git@master

wget https://gist.githubusercontent.com/ianporada/541721f439b0a30dfb1dc6b4db3d6946/raw/1cb4b597b9ef1ca9ca7f788dbd2a1a24cee2be44/bug_report_model.py
python bug_report_model.py
```

### Expected behavior

No NCCL error.

### Environment

* CUDA:
        - GPU:
                - NVIDIA Tesla P100-PCIE-16GB
                - NVIDIA Tesla P100-PCIE-16GB
                - NVIDIA Tesla P100-PCIE-16GB
                - NVIDIA Tesla P100-PCIE-16GB
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.20.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.0
        - pytorch-lightning: 1.4.0dev
        - tqdm:              4.61.1
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           # 51~18.04.1-Ubuntu SMP Fri Jun 4 15:21:28 UTC 2021

### Additional context

<details>
  <summary>The entire output for `NCCL_DEBUG=INFO  python bug_report_model.py`:</summary>

  ```
  /anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/metrics/__init__.py:43: LightningDeprecationWarning: `pytorch_lightning.metrics.*` module has been renamed to `torchmetrics.*` and split off to its own package (https://github.com/PyTorchLightning/metrics) since v1.3 and will be removed in v1.5
    rank_zero_deprecation(
  GPU available: True, used: True
  TPU available: False, using: 0 TPU cores
  IPU available: False, using: 0 IPUs
  /anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py:500: LightningDeprecationWarning: `trainer.fit(train_dataloader)` is deprecated in v1.4 and will be removed in v1.6. Use `trainer.fit(train_dataloaders)` instead. HINT: added 's'
    rank_zero_deprecation(
  initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/3
  initializing ddp: GLOBAL_RANK: 2, MEMBER: 3/3
  initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/3
  vm1:7389:7389 [0] NCCL INFO Bootstrap : Using [0]eth0:10.10.0.4<0> [1]veth2e03963:fe80::84ab:fbff:fe3e:a22%veth2e03963<0>
  vm1:7389:7389 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
  vm1:7389:7389 [0] NCCL INFO NET/IB : No device found.
  vm1:7389:7389 [0] NCCL INFO NET/Socket : Using [0]eth0:10.10.0.4<0> [1]veth2e03963:fe80::84ab:fbff:fe3e:a22%veth2e03963<0>
  vm1:7389:7389 [0] NCCL INFO Using network Socket
  NCCL version 2.7.8+cuda10.2
  vm1:7392:7392 [1] NCCL INFO Bootstrap : Using [0]eth0:10.10.0.4<0> [1]veth2e03963:fe80::84ab:fbff:fe3e:a22%veth2e03963<0>
  vm1:7392:7392 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
  vm1:7392:7392 [1] NCCL INFO NET/IB : No device found.
  vm1:7392:7392 [1] NCCL INFO NET/Socket : Using [0]eth0:10.10.0.4<0> [1]veth2e03963:fe80::84ab:fbff:fe3e:a22%veth2e03963<0>
  vm1:7392:7392 [1] NCCL INFO Using network Socket
  vm1:7403:7403 [2] NCCL INFO Bootstrap : Using [0]eth0:10.10.0.4<0> [1]veth2e03963:fe80::84ab:fbff:fe3e:a22%veth2e03963<0>
  vm1:7403:7403 [2] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
  vm1:7403:7403 [2] NCCL INFO NET/IB : No device found.
  vm1:7403:7403 [2] NCCL INFO NET/Socket : Using [0]eth0:10.10.0.4<0> [1]veth2e03963:fe80::84ab:fbff:fe3e:a22%veth2e03963<0>
  vm1:7403:7403 [2] NCCL INFO Using network Socket
  vm1:7392:7422 [1] NCCL INFO threadThresholds 8/8/64 | 24/8/64 | 8/8/64
  vm1:7389:7420 [0] NCCL INFO Channel 00/02 :    0   1   2
  vm1:7392:7422 [1] NCCL INFO Trees [0] 2/-1/-1->1->0|0->1->2/-1/-1 [1] 2/-1/-1->1->0|0->1->2/-1/-1
  vm1:7389:7420 [0] NCCL INFO Channel 01/02 :    0   1   2
  vm1:7403:7423 [2] NCCL INFO threadThresholds 8/8/64 | 24/8/64 | 8/8/64
  vm1:7392:7422 [1] NCCL INFO Setting affinity for GPU 1 to 0fff
  vm1:7403:7423 [2] NCCL INFO Trees [0] -1/-1/-1->2->1|1->2->-1/-1/-1 [1] -1/-1/-1->2->1|1->2->-1/-1/-1
  vm1:7403:7423 [2] NCCL INFO Setting affinity for GPU 2 to 0fff
  vm1:7389:7420 [0] NCCL INFO threadThresholds 8/8/64 | 24/8/64 | 8/8/64
  vm1:7389:7420 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1
  vm1:7389:7420 [0] NCCL INFO Setting affinity for GPU 0 to 0fff
  vm1:7389:7420 [0] NCCL INFO Channel 00 : 0[100000] -> 1[200000] via direct shared memory
  vm1:7403:7423 [2] NCCL INFO Channel 00 : 2[300000] -> 0[100000] via direct shared memory
  vm1:7392:7422 [1] NCCL INFO Channel 00 : 1[200000] -> 2[300000] via direct shared memory
  vm1:7403:7423 [2] NCCL INFO Channel 00 : 2[300000] -> 1[200000] via direct shared memory
  vm1:7392:7422 [1] NCCL INFO Channel 00 : 1[200000] -> 0[100000] via direct shared memory
  vm1:7389:7420 [0] NCCL INFO Channel 01 : 0[100000] -> 1[200000] via direct shared memory
  vm1:7403:7423 [2] NCCL INFO Channel 01 : 2[300000] -> 0[100000] via direct shared memory
  vm1:7392:7422 [1] NCCL INFO Channel 01 : 1[200000] -> 2[300000] via direct shared memory
  vm1:7403:7423 [2] NCCL INFO Channel 01 : 2[300000] -> 1[200000] via direct shared memory
  vm1:7392:7422 [1] NCCL INFO Channel 01 : 1[200000] -> 0[100000] via direct shared memory
  vm1:7389:7420 [0] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
  vm1:7389:7420 [0] NCCL INFO comm 0x7fdd80001060 rank 0 nranks 3 cudaDev 0 busId 100000 - Init COMPLETE
  vm1:7389:7389 [0] NCCL INFO Launch mode Parallel
  vm1:7403:7423 [2] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
  vm1:7392:7422 [1] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
  vm1:7403:7423 [2] NCCL INFO comm 0x7f9448001060 rank 2 nranks 3 cudaDev 2 busId 300000 - Init COMPLETE
  vm1:7392:7422 [1] NCCL INFO comm 0x7f4984001060 rank 1 nranks 3 cudaDev 1 busId 200000 - Init COMPLETE
  LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
  LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
  LOCAL_RANK: 2 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
  vm1:7392:7448 [2] NCCL INFO threadThresholds 8/8/64 | 24/8/64 | 8/8/64
  vm1:7403:7447 [3] NCCL INFO threadThresholds 8/8/64 | 24/8/64 | 8/8/64
  vm1:7389:7446 [1] NCCL INFO Channel 00/02 :    0   1   2
  vm1:7392:7448 [2] NCCL INFO Trees [0] 2/-1/-1->1->0|0->1->2/-1/-1 [1] 2/-1/-1->1->0|0->1->2/-1/-1
  vm1:7403:7447 [3] NCCL INFO Trees [0] -1/-1/-1->2->1|1->2->-1/-1/-1 [1] -1/-1/-1->2->1|1->2->-1/-1/-1
  vm1:7389:7446 [1] NCCL INFO Channel 01/02 :    0   1   2
  vm1:7403:7447 [3] NCCL INFO Setting affinity for GPU 3 to 0fff
  vm1:7392:7448 [2] NCCL INFO Setting affinity for GPU 2 to 0fff
  vm1:7389:7446 [1] NCCL INFO threadThresholds 8/8/64 | 24/8/64 | 8/8/64
  vm1:7389:7446 [1] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1
  vm1:7389:7446 [1] NCCL INFO Setting affinity for GPU 1 to 0fff
  vm1:7403:7447 [3] NCCL INFO Channel 00 : 2[400000] -> 0[200000] via direct shared memory
  vm1:7389:7446 [1] NCCL INFO Channel 00 : 0[200000] -> 1[300000] via direct shared memory
  vm1:7392:7448 [2] NCCL INFO Channel 00 : 1[300000] -> 2[400000] via direct shared memory
  vm1:7403:7447 [3] NCCL INFO Channel 00 : 2[400000] -> 1[300000] via direct shared memory
  vm1:7392:7448 [2] NCCL INFO Channel 00 : 1[300000] -> 0[200000] via direct shared memory
  vm1:7389:7446 [1] NCCL INFO Channel 01 : 0[200000] -> 1[300000] via direct shared memory
  vm1:7403:7447 [3] NCCL INFO Channel 01 : 2[400000] -> 0[200000] via direct shared memory
  vm1:7392:7448 [2] NCCL INFO Channel 01 : 1[300000] -> 2[400000] via direct shared memory
  vm1:7403:7447 [3] NCCL INFO Channel 01 : 2[400000] -> 1[300000] via direct shared memory
  vm1:7392:7448 [2] NCCL INFO Channel 01 : 1[300000] -> 0[200000] via direct shared memory
  vm1:7403:7447 [3] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
  vm1:7403:7447 [3] NCCL INFO comm 0x7f9420001060 rank 2 nranks 3 cudaDev 3 busId 400000 - Init COMPLETE
  vm1:7389:7446 [1] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
  vm1:7392:7448 [2] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
  vm1:7389:7446 [1] NCCL INFO comm 0x7fdd5c001060 rank 0 nranks 3 cudaDev 1 busId 200000 - Init COMPLETE
  vm1:7389:7389 [1] NCCL INFO Launch mode Parallel
  vm1:7392:7448 [2] NCCL INFO comm 0x7f4940001060 rank 1 nranks 3 cudaDev 2 busId 300000 - Init COMPLETE
  
  
  vm1:7392:7455 [1] init.cc:573 NCCL WARN Duplicate GPU detected : rank 2 and rank 1 both on CUDA device 200000
  vm1:7389:7454 [1] init.cc:573 NCCL WARN Duplicate GPU detected : rank 1 and rank 2 both on CUDA device 200000
  
  
  vm1:7389:7454 [1] NCCL INFO init.cc:840 -> 5
  vm1:7392:7456 [2] init.cc:573 NCCL WARN Duplicate GPU detected : rank 3 and rank 4 both on CUDA device 300000
  vm1:7403:7457 [2] init.cc:573 NCCL WARN Duplicate GPU detected : rank 4 and rank 3 both on CUDA device 300000
  vm1:7403:7457 [2] NCCL INFO init.cc:840 -> 5
  vm1:7389:7454 [1] NCCL INFO group.cc:73 -> 5 [Async thread]
  vm1:7392:7455 [1] NCCL INFO init.cc:840 -> 5
  vm1:7392:7456 [2] NCCL INFO init.cc:840 -> 5
  vm1:7403:7457 [2] NCCL INFO group.cc:73 -> 5 [Async thread]
  vm1:7392:7455 [1] NCCL INFO group.cc:73 -> 5 [Async thread]
  vm1:7392:7456 [2] NCCL INFO group.cc:73 -> 5 [Async thread]
  Traceback (most recent call last):
    File "/home/t-ianporada/pl-bert/bug_report_model.py", line 69, in <module>
      run()
    File "/home/t-ianporada/pl-bert/bug_report_model.py", line 64, in run
      trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 522, in fit
      self._run(model)
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 877, in _run
      self._pre_dispatch()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 905, in _pre_dispatch
      self.accelerator.pre_dispatch(self)
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 108, in pre_dispatch
      self.training_type_plugin.pre_dispatch()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 320, in pre_dispatch
      self.barrier()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 327, in barrier
      torch_distrib.barrier()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2420, in barrier
      work = default_pg.barrier(opts=opts)
  RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1614378062065/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, invalid usage, NCCL version 2.7.8
  ncclInvalidUsage: This usually reflects invalid usage of NCCL library (such as too many async ops, too many collectives at once, mixing streams in a group, etc).
  
  vm1:7389:7453 [0] include/socket.h:421 NCCL WARN Call to recv failed : Broken pipe
  vm1:7389:7453 [0] NCCL INFO include/socket.h:438 -> 2
  vm1:7389:7453 [0] NCCL INFO include/socket.h:444 -> 2
  vm1:7389:7453 [0] NCCL INFO bootstrap.cc:128 -> 2
  vm1:7389:7453 [0] NCCL INFO bootstrap.cc:351 -> 2
  vm1:7389:7453 [0] NCCL INFO init.cc:665 -> 2
  vm1:7389:7453 [0] NCCL INFO init.cc:840 -> 2
  vm1:7389:7453 [0] NCCL INFO group.cc:73 -> 2 [Async thread]
  
  vm1:7403:7458 [3] include/socket.h:421 NCCL WARN Call to recv failed : Broken pipe
  vm1:7403:7458 [3] NCCL INFO include/socket.h:438 -> 2
  vm1:7403:7458 [3] NCCL INFO include/socket.h:444 -> 2
  vm1:7403:7458 [3] NCCL INFO bootstrap.cc:128 -> 2
  vm1:7403:7458 [3] NCCL INFO bootstrap.cc:351 -> 2
  vm1:7403:7458 [3] NCCL INFO init.cc:665 -> 2
  vm1:7403:7458 [3] NCCL INFO init.cc:840 -> 2
  vm1:7403:7458 [3] NCCL INFO group.cc:73 -> 2 [Async thread]
  Traceback (most recent call last):
    File "bug_report_model.py", line 69, in <module>
  Traceback (most recent call last):
    File "/home/t-ianporada/pl-bert/bug_report_model.py", line 69, in <module>
      run()
    File "bug_report_model.py", line 64, in run
      trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 522, in fit
      run()
    File "/home/t-ianporada/pl-bert/bug_report_model.py", line 64, in run
      self._run(model)
      trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 877, in _run
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 522, in fit
      self._run(model)
      self._pre_dispatch()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 877, in _run
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 905, in _pre_dispatch
      self.accelerator.pre_dispatch(self)
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 108, in pre_dispatch
      self._pre_dispatch()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 905, in _pre_dispatch
      self.training_type_plugin.pre_dispatch()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 320, in pre_dispatch
      self.barrier()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 327, in barrier
      self.accelerator.pre_dispatch(self)
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 108, in pre_dispatch
      torch_distrib.barrier()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2420, in barrier
      self.training_type_plugin.pre_dispatch()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 320, in pre_dispatch
      self.barrier()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 327, in barrier
      torch_distrib.barrier()
    File "/anaconda/envs/bug_env/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2420, in barrier
      work = default_pg.barrier(opts=opts)
  RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1614378062065/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled system error, NCCL version 2.7.8
  ncclSystemError: System call (socket, malloc, munmap, etc) failed.
      work = default_pg.barrier(opts=opts)
  RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1614378062065/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled system error, NCCL version 2.7.8
  ncclSystemError: System call (socket, malloc, munmap, etc) failed.
  ```
</details>

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #8139

Can be reproduced on master with the following command:

```bash
NCCL_DEBUG=INFO python pl_examples/basic_examples/simple_image_classifier.py --trainer.gpus "1,3"  --trainer.accelerator  ddp
```

The issue appeared after #7202, where a barrier was introduced right after init_process group but before the ddp model gets configured with device_ids. When calling `torch.distributed.barrier` on a number of devices, it is assumed that it is affecting devices 0, 1, 2, etc. However, when we want to trainer on `gpus=[1,3]` for example that's not the case. 
The fix is to use the `device_ids` argument in the barrier call. I do not know why and how this works, don't ask me. But if you know why, let me know.

Partially related, which need to be informed about this fix:
- #4705
-  #7817

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
   **Note:** this cannot really be tested, needs a > 2 GPUs machine!
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -323,8 +323,13 @@ def pre_dispatch(self):
     def post_dispatch(self) -> None:
         self.cluster_environment.teardown()
 
-    def barrier(self, *args, **kwargs):
-        if torch_distrib.is_available() and torch_distrib.is_initialized():
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp.py:327]: reviewer: this is just refactor, not part of the fix

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -323,8 +323,13 @@ def pre_dispatch(self):
...
+    def barrier(self, *args, **kwargs) -> None:
+        if not torch_distrib.is_initialized():
+            return
+        device_ids = self.determine_ddp_device_ids()
+        if _TORCH_GREATER_EQUAL_1_8 and device_ids is not None:
+            torch_distrib.barrier(device_ids=device_ids)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp.py:331]: reviewer: this line is the fix.

**Review by justusschock — APPROVED** [review]

**Comment by kaushikb11** [comment]
@awaelchli Should we add a test for this?

**Review by kaushikb11 — APPROVED** [review]

**Comment by awaelchli** [comment]
@kaushikb11 I can add a test but it will require >3 gpus and will never run on our CI (which only has 2). We could then ask Adrian nicely to manually run this test on every ddp-related PR, what do you think? xD

**Comment by justusschock** [comment]
@awaelchli could we maybe mock the barrier call just to see if it is actually called with the devices instead of just the number?

**Review by tchaton — APPROVED** [review]
Wow, great fix ! Awesome work @awaelchli  :)

**Comment by awaelchli** [comment]
@kaushikb11 @justusschock thanks for your suggetions. I added a test, it's not particularly smart but it fails on master with NCCL error and passes here. Again, it will not run in our ci due to the gpu requirements, but I tested on another server by running the special test directly (master and this PR).

**Review by ananthsub — APPROVED** [review]
great catch @awaelchli !
I think there's a logline from torch.distributed that says when device ids is not explicitly passed, its inferred from the local rank, but I need to double check this.

**Comment by awaelchli** [comment]
@ananthsub This was also my suspicion, and I first tried to set `torch.cuda.set_device(local_rank)` before the first barrier call  without success.

In the [torch.distributed](https://pytorch.org/docs/stable/distributed.html) docs it is clearly written that it is the local rank by default for the `DistributedDataParallel` wrapper. 

> Please ensure that device_ids argument is set to be the only GPU device id that your code will be operating on. This is generally the local rank of the process. In other words, the device_ids needs to be [args.local_rank], and output_device needs to be args.local_rank in order to use this utility.

And it seems the barrier(device_ids=) need to match this. However, it is a bit a puzzle for me why the barrier needs this when other collective calls don't. I tried to set the device via

**Pull request merged**